File: leetcode/src/main/java/leetcode/array/FindPivotIndex.java
package leetcode.array;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.*;

/*
    https://leetcode.com/problems/find-pivot-index/

    Cho mảng số nguyên n phần tử, tìm vị trí i (đếm từ 1) chia mảng thành 2 phần

        arr[0] + arr[1] + ... + arr[i - 1] = arr[i + 1] + arr[i + 2] + ... + arr[n - 1]

    Nếu không có thì trả về -1.
 */
public class FindPivotIndex {

    // Nếu i = 0  thì leftSum = 0 do không có phần tử nào, tương tự cho rightSum khi i ở cuối
    public int pivotIndex(int[] nums) {
        int leftSum = 0;
        int totalSum = Arrays.stream(nums).sum();
        for (int i = 0; i < nums.length; i++) {
            int rightSum = totalSum - leftSum - nums[i];
            if (leftSum == rightSum) {
                return i;
            }
            leftSum += nums[i];
        }
        return -1;
    }

    @Test
    public void test() {
        assertEquals(3, pivotIndex(new int[]{1, 7, 3, 6, 5, 6}));
        assertEquals(-1, pivotIndex(new int[]{1, 2, 3}));
        assertEquals(0, pivotIndex(new int[]{2, 1, -1}));
    }

}

File: py-algorithm/arrays/find_pivot_index.py
import unittest

from math import gcd
from fractions import Fraction
from typing import List
from queue import PriorityQueue
from collections import Counter

"""
    https://leetcode.com/problems/find-pivot-index/

    Cho mảng số nguyên n phần tử, tìm vị trí i (đếm từ 1) chia mảng thành 2 phần

        arr[0] + arr[1] + ... + arr[i - 1] = arr[i + 1] + arr[i + 2] + ... + arr[n - 1]

    Nếu không có thì trả về -1.
"""


class Solution(unittest.TestCase):

    def pivotIndex(self, nums: List[int]) -> int:
        total = sum(nums)
        left_sum = 0
        for i in range(len(nums)):
            right_sum = total - left_sum - nums[i]
            if left_sum == right_sum:
                return i

            left_sum += nums[i]

        return -1

    def test(self):
        self.assertEqual(3, self.pivotIndex([1, 7, 3, 6, 5, 6]))
        self.assertEqual(-1, self.pivotIndex([1, 2, 3]))
        self.assertEqual(0, self.pivotIndex([2, 1, -1]))


if __name__ == '__main__':
    unittest.main()
